src/uucore/src/lib/features/i18n/mod.rs

```diff
@@ -0,0 +1,46 @@
+use std::{cmp::Ordering, sync::OnceLock};
+
+use icu::{
+    collator::{Collator, options::CollatorOptions},
+    locale::{Locale, locale},
+};
+
+fn get_collating_locale() -> &'static Locale {
+    static COLLATING_LOCALE: OnceLock<Locale> = OnceLock::new();
+
+    COLLATING_LOCALE.get_or_init(|| {
+        // Look at 3 environment variables in the following order
+        //
+        // 1. LC_ALL
+        // 2. LC_COLLATE
+        // 3. LANG
+        //
+        // Or fallback on Posix locale
+
+        let locale_var = std::env::var("LC_ALL")
+            .or_else(|_| std::env::var("LC_COLLATE"))
+            .or_else(|_| std::env::var("LANG"));
+
+        if let Ok(locale) = locale_var {
+            if let Some(simple) = locale.split(&['-', '@']).next() {
+                let bcp127 = simple.replace("_", "-");
+                if let Ok(locale) = Locale::try_from_str(&bcp127) {
+                    return locale;
+                }
+            }
+        }
+        // Default POSIX locale representing LC_ALL=C
+        locale!("en-US-posix")
+    })
+}
+
+pub fn locale_compare<T: AsRef<str>>(left: T, right: T) -> Ordering {
+    let collator = Collator::try_new(get_collating_locale().into(), CollatorOptions::default());
+
+    if let Ok(collator) = collator {
+        collator.compare_utf8(left.as_ref().as_bytes(), right.as_ref().as_bytes())
```

Comment:
Since this function already has arguments that have been validated to be well-formed UTF-8 (`AsRef<str>`), using `compare_utf8` is a slight pessimization and it makes sense to use `compare`.

`compare_utf8` is for the case where you have something that's meant to be UTF-8 in `&[u8]` but not validated as `&str`.